src/resource.ts

```diff
@@ -0,0 +1,207 @@
+import axios from "axios";
+import moment from "moment";
+import BlueButton from "./index";
+
+// initInterval in milli-seconds
+export const retrySettings = {
+  initInterval: 5000,
+  maxAttempts: 3,
+  backOffExpr: "interval * ( 2 ** i )",
+  retryableCodes: [500, 502, 503, 504],
+  retryableFhirPath: [
+    "/fhir/Patient/",
+    "/fhir/Coverage/",
+    "/fhir/ExplanationOfBenefit/",
+    "/connect/userinfo",
+  ],
+};
+
+export enum FhirResourceType {
+  Patient = "Patient",
+  Coverage = "Coverage",
+  Profile = "Profile",
+  ExplanationOfBenefit = "ExplanationOfBenefit",
+}
+
+export class AuthorizationToken {
+  public accessToken: string;
+  public expiresIn: number;
+  public expiresAt: number;
+  public tokenType: string;
+  public scope: [string];
+  public refreshToken: string;
+  public patient: string;
+
+  constructor(authToken: any) {
+    this.accessToken = authToken.access_token;
+    this.expiresIn = authToken.expires_in; // in seconds
+    this.expiresAt = authToken.expires_at
+      ? authToken.expires_at
+      : moment()
+          .add(this.expiresIn * 1000)
+          .valueOf();
+    this.patient = authToken.patient;
+    this.refreshToken = authToken.refresh_token;
+    this.scope = authToken.scope;
+    this.tokenType = authToken.token_type;
+  }
+
+  isExpired(): boolean {
+    return moment(this.expiresAt).isBefore(moment());
+  }
+
+  validate(): void {
+    if (
+      !(
+        this.accessToken &&
+        this.expiresIn &&
+        this.expiresAt &&
+        this.patient &&
+        this.refreshToken &&
+        this.scope &&
+        this.tokenType
+      )
+    ) {
+      throw new Error("Invalid authorization token.");
+    }
+  }
+}
+
+function sleep(time: number) {
+  return new Promise((resolve) => {
+    setTimeout(resolve, time);
+  });
+}
+
+function isRetryable(error: any) {
+  let retryOK = false;
+
+  if (
+    error.response &&
+    retrySettings.retryableCodes.includes(error.response.status)
+  ) {
+    if (error.request && error.request.path) {
+      retryOK = endsWith(error.request.path);
+    }
+  }
+  return retryOK;
+}
+
+function endsWith(path: string) {
+  let matched = false;
+  for (let i = 0; i < retrySettings.retryableFhirPath.length; i++) {
+    if (path.endsWith(retrySettings.retryableFhirPath[i])) {
+      matched = true;
+      break;
+    }
+  }
+  return matched;
+}
+
+async function doRetry(fhirUrl: string, config: any) {
+  const interval = retrySettings.initInterval;
+  const maxAttempts = retrySettings.maxAttempts;
+
+  let resp = null;
+
+  for (let i = 0; i < maxAttempts; i += 1) {
+    const waitInMilliSec = eval(retrySettings.backOffExpr);
```

Comment:
I generally try to avoid eval based functions. If someone is able to override the `retrySettings.backOffExpr` with bad code we would be in trouble. I don't know if it's strictly required here either.